cstrml/market/src/lib.rs

```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
                     T::Payment::pay_sorder(order_id);
                 }
 
-                // TODO: add market slashing here
-                // TODO: Record `failed_count` from `Success` to `Failed`
-                // TODO: Record `failed_duration` from `Failed` to `Success`
-
                 // 2. Update storage order
                 <StorageOrders<T>>::insert(order_id, so);
             }
         }
     }
 
+    pub fn update_slash_record(order_id: &<T as system::Trait>::Hash) {
+        // 1. Update record's count
+        let so = Self::storage_orders(order_id).unwrap();
+        <OrderSlashRecords<T>>::mutate(order_id, |status_count| {
+            if let Some(sc) = status_count {
+                if so.status == OrderStatus::Success {
+                    sc.success += 1;
+                } else if so.status == OrderStatus::Failed {
+                    sc.failed += 1;
+                } else {}
+            }
+
+        });
+
+        // 2. Maybe do slash
+        Self::maybe_slash_provider(&order_id, &so);
+    }
+
+    pub fn maybe_slash_provider(order_id: &T::Hash, so: &StorageOrder<T::AccountId, BalanceOf<T>>) {
+        let mut status_count = Self::order_slash_records(order_id).unwrap();
+        let slash_duration = T::SlashingDuration::get();
+        if status_count.success + status_count.failed >= slash_duration && status_count.can_slash {
```

Comment:
Is there a simple way to judge the punishing condition?
Like recording the `current_era`?

Reply:
each order has its own check and cannot have a current_era.